### ISSUE TITLE
SDDM 0.20 uses random TTY, breaks autostart programs needing X server

### ISSUE THREAD

**Issue by beermad** [sddm/sddm#1777]
Since upgrading to SDDM 20.0, my X server no longer starts reliably on TTY1, instead it seems to be fairly random, though it _usually_ opens on TTY2.

In principle this shouldn't be a problem in itself, but unfortunately it has very problematic side-effects. I have a number of programs which I launch via KDE autostart options. Some of these need to open GUI windows (things like Firefox & Thunderbird, for example). 

Other users on Reddit and the Manjaro forum also report that cron jobs and/or systemd timers which should open GUI windows are failing. 

All failures are accompanied by the following messages:
```
Invalid MIT-MAGIC-COOKIE-1 key 
Cannot open display ":0.0"
```
[Edit]: the magic cookie message may be a red herring. But "Cannot open display" is the important message".

Non-startup jobs can be fixed by issuing:
`xhost +`

Startup jobs _will_ work OK if I log out then log back in again, so it seems the problem is that somehow the change of TTY is causing an INITIAL problem opening the X display. I suspect other people's problem with cron/systemd jobs would also get fixed by that.

I have downgraded SDDM to the 0.19 and this has solved my problems, which proves the new version to be responsible.

**Comment by MrLavender** [comment]
This line

https://github.com/sddm/sddm/blob/c813b2f56c28e909c17e84c87430ff2885a2f812/src/daemon/Display.cpp#L119

introduced in https://github.com/sddm/sddm/commit/ebe6110bd2bb5047ca09d4446fe739da468086e1 means it can never start on TTY1. Changing it to use `fetchAvailableVt()` instead restores the old behaviour.

**DISCLAIMER** I have no idea what the original reasons were for using `VirtualTerminal::setUpNewVt()` there instead of `fetchAvailableVt()`. And I have only tested this change with X11 on my system. It Works For Me<sup>TM</sup>.

**Comment by davidedmundson** [comment]
It's the display something other than :0.0 in the broken case?

I strongly suspect this isn't about TTYS but about timing of the greeters x11 session and the magic cookie is the relevant part.

**Comment by kraithus** [comment]
I'm having similar problem. Cron jobs using xterm not working anymore since updating to this version. 

Checking systemctl status cronie I get:

```
Authorization required, but no authorization protocol specified and

xterm: Xt error: Can't open display: :0
```

`xhost +SI:localuser:username` on each startup seems to fix the issue

**Comment by Vogtinator** [comment]
> Other users on Reddit and the Manjaro forum also report that cron jobs and/or systemd timers which should open GUI windows are failing.

System wide ones? Probably because they use a hardcoded `$DISPLAY` or `$XAUTHORITY`, which is broken by design.

**Comment by beermad** [comment]
> Probably because they use a hardcoded `$DISPLAY` or `$XAUTHORITY`

Perhaps a quote from Linus may be significant here: "We do not break user-space. If a change to the Linux kernel breaks a program it's automatically a kernel bug which needs to be fixed".

So when (probably large) numbers of users are following common practice and that's been working for years (in my case, decades), and what they're doing no longer works due to a change in SDDM, isn't it fair to say that the problem is the change rather than the users' settings?

**Comment by Vogtinator** [comment]
Hardcoding `$DISPLAY` was always wrong, it might randomly fail.

The change for `$XAUTHORITY` was needed to address important bugs.

We don't just break stuff to annoy users, there are valid reasons behind it.

**Comment by beermad** [comment]
> Hardcoding `$DISPLAY` was always wrong, it might randomly fail.

It's certainly something I've often seen recommended and is very common practice. You're going to end up with an awful lot of people with broken computers when this release rolls out to slower-updating distros.

And it's curious that if I log out then back in again, background-launched GUI programs _do_ run successfully. It seems that something SDDM is doing (or not doing) when first launched after a boot is causing the problem.

**Comment by Vogtinator** [comment]
> > Hardcoding `$DISPLAY` was always wrong, it might randomly fail.
> 
> It's certainly something I've often seen recommended and is very common practice. You're going to end up with an awful lot of people with broken computers when this release rolls out to slower-updating distros.

The `$DISPLAY` behaviour didn't actually change in SDDM 0.20, unless you use `DisplayServer=x11-user`.

> And it's curious that if I log out then back in again, background-launched GUI programs _do_ run successfully. It seems that something SDDM is doing (or not doing) when first launched after a boot is causing the problem.

What do you mean by "background-launched GUI programs"? Autostart applications get the environment from the DE.

> [Edit]: the magic cookie message may be a red herring. But "Cannot open display" is the important message".

It's the other way around: The magic cookie message is the cause, the open failure is the outcome.

**Comment by beermad** [comment]
OK, I've now tested the "hard coded $DISPLAY" theory. Everywhere I was setting $DISPLAY is commented out and I re-installed 0.20 then rebooted.

Result: on the first login, GUI programs such as Firefox _still_ fail to start. At the start of the script which runs Firefox, I've added "echo $DISPLAY" and this shows that no DISPLAY variable is set. But when the script attempts to launch Firefox it throws an error:
`Error: no DISPLAY environment variable specified`

When I log out and back in again, everything starts as it was doing in the past. The script starting Firefox now reports that when it starts, $DISPLAY is set to `:0`.

I'm now going to downgrade to 0.19 again to see what happens with no DISPLAY set in my startup scripts.

Note: I'm not setting $XAUTHORITY in any of my scripts.

**Comment by beermad** [comment]
Having gone back to 0.19 and rebooted, with the DISPLAY variable not being set, I find that none of my GUI programs will run _without_ DISPLAY being set. Even after logging off and logging back in again. I'm not entirely sure that proves anything useful beyond what my last post said, except that one way or another I still need to be setting $DISPLAY for autostarts.

**Comment by Vogtinator** [comment]
How do you autostart those applications? If `$DISPLAY` is not set that means something is **really** broken with the DE setup.

**Comment by beermad** [comment]
> How do you autostart those applications? If `$DISPLAY` is not set that means something is **really** broken with the DE setup.

A wrapper script is called as a KDE startup program so all the things I want to launch at startup run automatically. This script calls various others, one of which for example, starts Firefox, as previously mentioned.

So I assume KDE ought to be passing the necessary environment variables, though evidently this isn't happening.

**Comment by Vogtinator** [comment]
Which wrapper script and how did you set up the startup program? Do you use systemdBoot (default since Plasma 5.26)?

**Comment by medmedin2014** [comment]
On Manjaro KDE, my Wayland session boots using tty1, but when I press CTRL+ALT+F2 I don’t get the textual virtual terminal, I see only a frozen SDDM screen, while with F3,F4,F5,F6 it works fine.

**Comment by Vogtinator** [comment]
> On Manjaro KDE, my Wayland session boots using tty1, but when I press CTRL+ALT+F2 I don’t get the textual virtual terminal, I see only a frozen SDDM screen, while with F3,F4,F5,F6 it works fine.

Yeah, the greeter is in a zombie state while a Wayland session is open. This should not be an issue in practice.
I'd expect SDDM to be on tty2 though and the wayland session on tty3. That tty1 is available for the wayland session but not the greeter one is odd. Maybe systemd has ownership of the VT only until booting finishes.

**Comment by beermad** [comment]
> Which wrapper script and how did you set up the startup program? Do you use systemdBoot (default since Plasma 5.26)?

The wrapper script fires up various things I need started with my session, so it's just one I've put together for my purposes (I could post it here but it probably wouldn't be helpful as there's a lot there that isn't relevant).
It's set up in KDE systemsettings: Startup and Shutdown=>Autostart=>Login scripts.

This is a machine that was built before systemdBoot was default, so it still uses GRUB (and looking at guides to migrating, I'm not going to be able to do that without seriously messing about with my disc partitioning).

**Comment by Vogtinator** [comment]
> > Which wrapper script and how did you set up the startup program? Do you use systemdBoot (default since Plasma 5.26)?
> 
> The wrapper script fires up various things I need started with my session, so it's just one I've put together for my purposes (I could post it here but it probably wouldn't be helpful as there's a lot there that isn't relevant). It's set up in KDE systemsettings: Startup and Shutdown=>Autostart=>Login scripts.

Ok. Maybe those scripts are somehow run in the wrong context, but at least here it works. This script got converted into a `.desktop` file by some Plasma update and the xdg-autostart-generator generated `app-foobar.sh@autostart.service`.

> This is a machine that was built before systemdBoot was default, so it still uses GRUB (and looking at guides to migrating, I'm not going to be able to do that without seriously messing about with my disc partitioning).

Note that this is about the systemdBoot feature in Plasma, not the systemd-boot bootloader.

**Comment by beermad** [comment]
> Note that this is about the systemdBoot feature in Plasma, not the systemd-boot bootloader.

Ah! Sorry, misunderstood you there.
Unless I've found the wrong clues with a web search, it appears I _am_ using that.
```
$ kreadconfig5 --file startkderc --group General --key systemdBoot
true
```
`systemctl  --user list-units|grep -i autostart` shows a service called app-startup@autostart.service and that appears to be running my startup script.

**Comment by Vogtinator** [comment]
Sounds good. If you restart that service, does it work as expected? If yes, it might be a race condition. If not, check `systemctl --user show-environment`.

**Comment by beermad** [comment]
> Sounds good. If you restart that service, does it work as expected? If yes, it might be a race condition. If not, check `systemctl --user show-environment`.

If I restart the service, it _does_ run as expected. A little bit more digging finds something that may be significant.

`systemctl --user show-environment` from a terminal shows $DISPLAY as :0. I put that command in my startup script then rebooted. Looking at the output from the script when it was automatically started, $DISPLAY _hadn't_ been set. So my suspicion is that for some reason that variable isn't being passed into the environment. Though my ignorance of the way SDDM and systemd interact may mean this is utter rubbish.

**Comment by Vogtinator** [comment]
> So my suspicion is that for some reason that variable isn't being passed into the environment. Though my ignorance of the way SDDM and systemd interact may mean this is utter rubbish.

No, that's probably the cause. This shouldn't happen though, the variables should be set by `pam_systemd` already and even extended by Plasma itself during start, as it does the equivalent of `dbus-update-activation-environment --systemd` before running those scripts.

Are other variables (like `XAUTHORITY`, `XDG_SESSION_TYPE`, etc.) also missing?

**Comment by beermad** [comment]
> Are other variables (like `XAUTHORITY`, `XDG_SESSION_TYPE`, etc.) also missing?

They're also missing. Full output of systemctl --user show-environment attached.
[environment.log](https://github.com/sddm/sddm/files/12266758/environment.log)

Though they _are_ set in a shell session  after logging in.
 I wonder if Plasma might be launching startup scripts before SDDM has had chance to set up the environment?

**Comment by Vogtinator** [comment]
> Yhey're also missing. Full output of systemctl --user show-environment attached.
[environment.log](https://github.com/sddm/sddm/files/12266758/environment.log)

That is extremely empty, it does not even have `XDG_SESSION_ID` which is set up by `pam_systemd` itself! I suspect something might delete those variables somehow. If you start a non-Plasma session like IceWM or something similarly basic, is that included? If so, it might be that the initial sync from Plasma's env deletes them.

> I wonder if Plasma might be launching startup scripts before SDDM has had chance to set up the environment?

No, starting the session process is the very last thing done by SDDM.

**Comment by beermad** [comment]
> If you start a non-Plasma session like IceWM or something similarly basic, is that included? If so, it might be that the initial sync from Plasma's env deletes them.

I don't have any other environments installed at the moment, so I'll have to set one up on my "rescue" partition. Might take a couple of days before I've got a long enough window to give it a try.

**Comment by Vogtinator** [comment]
You could also create a new file in `/usr/share/xsessions/` to just start e.g. xterm.

**Comment by beermad** [comment]
> You could also create a new file in `/usr/share/xsessions/` to just start e.g. xterm.

OK, I'll look into that and report back.

**Comment by davidedmundson** [comment]
>I suspect something might delete those variables somehow.

I know what's causing this. KDE's env sync is a two step process.

First we drop all old env vars from systemd's knowledge (see startplasma.cpp dropSessionVarsFromSystemdEnvironment)
Then we import all relevant new ones as a separate step.

The first step (clearing) is working, the second step (importing the new set) is not.

There used to be a quirk whereby systemd would reject the entire updateEnv command if a single key-value pair didn't match their arbitrary security check. We have a bunch of guards in place, but clearly not enough.

**Comment by davidedmundson** [comment]
Can you attach output of `journalctl --user -b`  after a login to plasma with the failing state.

I suspect we'll see a line starting:

 "Failed to unset systemd environment variables:" << reply.errorName() << reply.errorMessage();

**Comment by Vogtinator** [comment]
That's my suspicion as well. If it's true then with the plain xterm the session vars should be there.

The other question is why it starts working at some later point. I suspect something else triggers a different sync later on that succeeds.

**Comment by beermad** [comment]
> Can you attach output of `journalctl --user -b` after a login to plasma with the failing state.

Unfortunately it's not playing. All I get is
`-- No entries --`
There's something odd going on with my user journal and it only seems to actually start logging at some random time after a boot. I may have a misconfiguration in journald.conf. Something else for me to try to sort out...

**Comment by beermad** [comment]
> That is extremely empty, it does not even have `XDG_SESSION_ID` which is set up by `pam_systemd` itself! I suspect something might delete those variables somehow. If you start a non-Plasma session like IceWM or something similarly basic, is that included? If so, it might be that the initial sync from Plasma's env deletes them.

Output from systemctl --user show-environment from an xterm session attached.
[env.log](https://github.com/sddm/sddm/files/12269901/env.log)
It's setting $DISPLAY and $XAUTHORITY but not $XDG_SESSION_ID

**Comment by Vogtinator** [comment]
Ok, so that confirms those variables are indeed unset by Plasma. Why `$XDG_SESSION_ID` is missing I can't say, apparently `pam_systemd` doesn't forward the variables that it set itself to the systemd instance it started itself...

**Comment by breakingspell** [comment]
This issue is still present in latest SDDM. 

Rather than autostart, my use case is two X sessions (personal on TTY1, work on TTY2, simply set by login order). This allows me to use Ctrl+Alt+F1 and F2 to switch between sessions quickly rather than using the DE userswitch/screenlock. 
Since 20, my personal session may be on TTY1 or TTY2, while the work session can be anywhere from TTY5 to TTY8. I can confirm this on a fresh system with no hardcoded session variables. 

I tried a quick patch based on this [user comment](https://github.com/sddm/sddm/issues/1777#issuecomment-1659510268), but this seems to break creating additional sessions. Rolling back to SDDM 19.0 had been my workaround for a while, but now Plasma 6 expects SDDM 20.

**Comment by Vogtinator** [comment]
> Since 20, my personal session may be on TTY1 or TTY2, while the work session can be anywhere from TTY5 to TTY8. I can confirm this on a fresh system with no hardcoded session variables.

Might be fixed by https://github.com/sddm/sddm/pull/1896, depending on your tty setup.
